File: day-9/python/main.py
import sys
import numpy as np


def parse_args():
    if len(sys.argv) != 2:
        sys.exit('I need a file')
    return sys.argv[1]


def parse_input(filename):
    data = []
    with open(filename) as fd:
        for line in fd:
            s = line.split()
            data.append([s[0], int(s[1])])
    return data


def move(m, i, j):
    if m == 'R':
        return (i, j + 1)
    if m == 'L':
        return (i, j - 1)
    if m == 'U':
        return (i + 1, j)
    return (i - 1, j)


def follow_dir(h, t):
    if (h - t) == 2:
        return t + 1
    if (t - h) == 2:
        return t - 1
    return t


def follow(hi, hj, ti, tj):
    if abs(hi - ti) < 2 and abs(hj -tj) < 2:
        return (ti, tj)

    nti = follow_dir(hi, ti)
    ntj = follow_dir(hj, tj)

    if abs(hi - ti) == 2 and abs(hj - tj) == 1:
        ntj = hj
    if abs(hi - ti) == 1 and abs(hj - tj) == 2:
        nti = hi

    return nti, ntj


def problem_1(data):
    ti, tj = 0, 0
    hi, hj = 0, 0
    pos = set()
    for m, s in data:
        for t in range(s):
            hi, hj = move(m, hi, hj)
            ti, tj = follow(hi, hj, ti, tj)
            pos.add((ti, tj))
    return len(pos)


def problem_2(data):
    nodes = []
    for i in range(10):
        nodes.append([0, 0])
    last_pos = set()
    for m, s in data:
        for t in range(s):
            nodes[0] = move(m, nodes[0][0], nodes[0][1])
            for j in range(1, len(nodes)):
                nodes[j] = follow(nodes[j - 1][0], nodes[j - 1][1],
                                  nodes[j][0], nodes[j][1])
                if j == len(nodes) - 1:
                    last_pos.add(nodes[j])
    return len(last_pos)


def main():
    filename = parse_args()
    data = parse_input(filename)
    print("Day 9, problem 1: ", problem_1(data))
    print("Day 9, problem 2: ", problem_2(data))


if __name__ == '__main__':
    main()

File: day-9/rust/src/main.rs
use std::env;
use std::fs::File;
use std::io::{self, BufRead};
use std::process;
use std::collections::HashSet;

fn parse_args() -> String {
    let mut args = env::args();
    if args.len() != 2 {
        eprintln!("Invalid arguments: I need the input file and nothing else");
        process::exit(1);
    }
    args.nth(1).unwrap()
}

fn create_reader(filename: &String) -> impl Iterator<Item = (char, u32)> {
    let file = File::open(&filename).unwrap_or_else(|_| {
        eprintln!("Cannot open file: {}", &filename);
        process::exit(1);
    });
    io::BufReader::new(file)
        .lines()
        .map(|l| l.unwrap()
                  .split_whitespace()
                  .map(|x| x.to_string())
                  .collect::<Vec<String>>())
        .map(|x| {
            let mut it = x.iter();
            let m = it.next().unwrap().chars().next().unwrap();
            let q = it.next().unwrap().parse::<u32>().unwrap();
            (m, q)
        })
}

fn head_move(m: char, p: &mut [i32; 2]) {
    if m == 'R' {
        p[1] += 1;
    } else if m == 'L' {
        p[1] -= 1;
    } else if m == 'U' {
        p[0] += 1;
    } else {
        p[0] -= 1;
    }
}

fn follow_dir(h: i32, t: i32) -> i32 {
    if (h - t) == 2 {
        return t + 1;
    }
    if (t - h) == 2 {
        return t - 1;
    }
    return t
}

fn follow(prev: [i32; 2], node: &mut [i32; 2]) {
    if (prev[0] - node[0]).abs() < 2 && (prev[1] - node[1]).abs() < 2 {
        return;
    }
    let mut nti = follow_dir(prev[0], node[0]);
    let mut ntj = follow_dir(prev[1], node[1]);

    if (prev[0] - node[0]).abs() == 2 && (prev[1] - node[1]).abs() == 1 {
        ntj = prev[1];
    }
    if (prev[0] - node[0]).abs() == 1 && (prev[1] - node[1]).abs() == 2 {
        nti = prev[0];
    }
    node[0] = nti;
    node[1] = ntj;
}

fn simulate(moves: impl Iterator<Item = (char, u32)>, nodes: usize) -> usize {
    let mut p = vec![[0 as i32; 2]; nodes];
    let mut tail_pos = HashSet::<[i32; 2]>::new();
    for (m, q) in moves {
        for _ in 0..q {
            head_move(m, &mut p[0]);
            for j in 1..nodes {
                follow(p[j - 1], &mut p[j]);
                if j == nodes - 1 {
                    tail_pos.insert(p[j]);
                }
            }
        }
    }
    tail_pos.len()
}

fn main() {
    let filename = parse_args();

    println!("Day 9, problem 1: {}", simulate(create_reader(&filename), 2));
    println!("Day 9, problem 2: {}", simulate(create_reader(&filename), 10));
}
